core/drivers/versal_pm.c

```diff
@@ -0,0 +1,220 @@
+#define PDI_SRC_SMAP		0xA
+#define PDI_SRC_PCIE		0xB
+#define PDI_SRC_SD1_LS		0xE
+#define PDI_SRC_DDR		0xF
+
+struct versal_sip_payload {
+	uint32_t data[PAYLOAD_ARG_CNT];
+};
+
+static uint32_t versal_sip_call(uint32_t pm_api_id, uint32_t arg0,
+				uint32_t arg1, uint32_t arg2, uint32_t arg3,
+				struct versal_sip_payload *payload)
+{
+	struct thread_smc_args args = {
+		.a0 = pm_api_id > PM_LOAD_PDI ?
+		      pm_api_id : VERSAL_SIP_SVC | pm_api_id,
+		.a1 = reg_pair_to_64(arg1, arg0),
+		.a2 = reg_pair_to_64(arg3, arg2),
+	};
+
+	thread_smccc(&args);
+
+	if (payload) {
+		reg_pair_from_64(args.a0, &payload->data[1], &payload->data[0]);
+		reg_pair_from_64(args.a1, &payload->data[3], &payload->data[2]);
+		reg_pair_from_64(args.a2, &payload->data[5], &payload->data[4]);
+		reg_pair_from_64(args.a3, &payload->data[7], &payload->data[6]);
+	}
+
+	/* allow the PLM to output its debug information */
+	if (IS_ENABLED(CFG_VERSAL_TRACE_PLM))
+		mdelay(500);
+
+	return args.a0;
+}
+
+/* SIP call to program the FPGA has been obsoleted, use the PLM */
+TEE_Result versal_write_fpga(paddr_t pa)
+{
+	struct ipi_cmd cmd = { };
+
+	cmd.data[0] = PM_LOAD_PDI;
+	cmd.data[1] = PDI_SRC_DDR;
+	cmd.data[2] = pa >> 32;
+	cmd.data[3] = pa;
+	if (versal_mbox_notify(&cmd, NULL, NULL))
+		return TEE_ERROR_GENERIC;
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result versal_soc_version(uint8_t *version)
+{
+	struct versal_sip_payload payload = { };
+	const uint32_t version_shift = 12;
+	uint32_t res = 0;
+
+	if (!version)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = versal_sip_call(PM_GET_CHIPID, 0, 0, 0, 0, &payload);
+	if (res) {
+		EMSG("Failed to retrieve version");
+		return TEE_ERROR_GENERIC;
+	}
+
+	*version = payload.data[2] >> version_shift;
+
+	return TEE_SUCCESS;
+}
+
+static bool uuid_is_versal_pm(void)
+{
+	struct versal_sip_payload p = { };
+
+	versal_sip_call(VERSAL_SIP_SVC_UID, 0, 0, 0, 0, &p);
+
+	if (p.data[0] == VERSAL_SIP_UID_0 && p.data[2] == VERSAL_SIP_UID_1 &&
+	    p.data[4] == VERSAL_SIP_UID_2 && p.data[6] == VERSAL_SIP_UID_3)
+		return true;
+
+	return false;
+}
+
+static TEE_Result versal_check_pm_abi(void)
+{
+	struct versal_sip_payload p = { };
+	struct ipi_cmd cmd = { };
+	struct ipi_cmd rsp = { };
+	uint32_t major = 0;
+	uint32_t minor = 0;
+
+	if (!uuid_is_versal_pm()) {
+		EMSG("Invalid SiP Service");
+		return TEE_ERROR_GENERIC;
+	}
+
+	if (versal_sip_call(VERSAL_SIP_SVC_VERSION, 0, 0, 0, 0, &p))
+		return TEE_ERROR_GENERIC;
+
+	major = p.data[0];
+	minor = p.data[2];
+	if (major != VERSAL_SIP_MAJOR || minor < VERSAL_SIP_MINOR) {
+		EMSG("Invalid SiP version: Major %d, Minor %d", major, minor);
+		return TEE_ERROR_GENERIC;
+	}
+
+	cmd.data[0] = PM_API_ID(PM_GET_API_VERSION);
+	if (versal_mbox_notify(&cmd, &rsp, NULL))
+		return TEE_ERROR_GENERIC;
+
+	major = rsp.data[1] & 0xFFFF;
+	minor = rsp.data[1] >> 16;
+	if (major != VERSAL_PM_MAJOR || minor < VERSAL_PM_MINOR) {
+		EMSG("Invalid PM version: Major %d, Minor %d", major, minor);
```

Comment:
s/%d/%"PRId32"/
or more simply define `major` and `minor` with type `unsigned int`.